lib/msal-browser/src/app/PublicClientApplication.ts

```diff
@@ -187,9 +193,64 @@ export class PublicClientApplication extends ClientApplication implements IPubli
-            });
+
+            const acquireTokenByRefreshToken = (silentRequest: CommonSilentFlowRequest, silentTokenRetrievalStrategy: SilentTokenRetrievalStrategy | undefined): Promise<AuthenticationResult> => {
+                const atbrtMeasurement = this.performanceClient.startMeasurement(PerformanceEvents.AcquireTokenByRefreshToken, request.correlationId);
+
+                /*
+                 * do not attempt to get the access token via the refresh token (just go to the iframe) if the
+                 * silentTokenRetrievalStrategy is set to NetworkOnly
+                 */
+                if (silentTokenRetrievalStrategy === SilentTokenRetrievalStrategy.NetworkOnly) {
+                    const networkOnly = true;
+                    return this.acquireTokenBySilentIframe(undefined, silentRequest, atbrtMeasurement, networkOnly);
+                }
+
+                // otherwise, attempt to get the access token via the refresh token before going to the iframe
+                return this.acquireTokenByRefreshToken(silentRequest, atbrtMeasurement).catch(async (error) => {
+                    /*
+                     * only attempt to get the access token via the iframe (network) if the
+                     * silentTokenRetrievalStrategy is not set to RefreshTokenOnly or CacheOrRefreshToken
+                     */
+                    if ((silentTokenRetrievalStrategy !== SilentTokenRetrievalStrategy.RefreshTokenOnly)
+                        && (silentTokenRetrievalStrategy !== SilentTokenRetrievalStrategy.CacheOrRefreshToken)) {
+                        const networkOnly = false;
+                        return this.acquireTokenBySilentIframe(error, silentRequest, atbrtMeasurement, networkOnly);
+                    } else {
+                        this.logger.info("SilentTokenRetrievalStrategy set to not renew expired refresh token.");
+                        atbrtMeasurement.endMeasurement({
+                            success: false
+                        });
+                        throw error;
+                    }
+                });
+            };
+
+            /*
+             * do not look in the cache (just go to the network) if the silentTokenRetrievalStrategy
+             * is set to one of the network-only options
+             */
+            if ((request.silentTokenRetrievalStrategy === SilentTokenRetrievalStrategy.RefreshTokenOnly)
+                || (request.silentTokenRetrievalStrategy === SilentTokenRetrievalStrategy.NetworkWithRefreshToken)
+                || (request.silentTokenRetrievalStrategy === SilentTokenRetrievalStrategy.NetworkOnly)) {
+                this.logger.info("SilentCacheClient:acquireToken - Skipping cache because SilentTokenRetrievalStrategy is set to one of the network-only options.");
+                
+                result = acquireTokenByRefreshToken(silentRequest, request.silentTokenRetrievalStrategy);
```

Comment:
Why are both parameters needed? `silentRequest` should have `silentTokenRetrievalStrategy` as a property.

Reply:
Ah, `silentRequest` is `CommonSilentFlowRequest`, which doesn't have `silentTokenRetrievalStrategy`. Will think about how to best handle this.

Reply:
Yep, that's also why I can't use the enum in PerformanceEvent.ts like you suggested above.